app/org/sagebionetworks/bridge/models/schedules/TaskScheduler.java

```diff
@@ -51,43 +41,46 @@ protected DateTime getScheduledTimeBasedOnEvent(Schedule schedule, Map<String, D
         return eventTime;
     }
     
-    protected void addTaskForEachTime(List<Task> tasks, DateTime scheduledTime) {
+    protected void addTaskForEachTime(List<Task> tasks, ScheduleContext context, DateTime scheduledTime) {
         if (schedule.getTimes().isEmpty()) {
             // We're using whatever hour/minute/seconds were in the original event.
-            addTaskForEachActivityAtTime(tasks, scheduledTime);
+            addTaskForEachActivityAtTime(tasks, context, scheduledTime);
         } else {
             for (LocalTime time : schedule.getTimes()) {
                 scheduledTime = new DateTime(scheduledTime).withTime(time);
-                addTaskForEachActivityAtTime(tasks, scheduledTime);
+                addTaskForEachActivityAtTime(tasks, context, scheduledTime);
             }
         }
     }
     
-    private void addTaskForEachActivityAtTime(List<Task> tasks, DateTime scheduledTime) {
+    private void addTaskForEachActivityAtTime(List<Task> tasks, ScheduleContext context, DateTime scheduledTime) {
         // If this time point is outside of the schedule's active window, skip it.
         if (isInWindow(schedule, scheduledTime)) {
             // As long at the tasks are not already expired, add them.
-            DateTime expiresOn = getExpiresOn(scheduledTime, schedule);
-            if (expiresOn == null || expiresOn.isAfter(now)) {
+            DateTime expiresOn = getExpiresOn(schedule, scheduledTime);
+            if (expiresOn == null || expiresOn.isAfter(context.getNow())) {
                 for (Activity activity : schedule.getActivities()) {
                     DynamoTask task = new DynamoTask();
-                    task.setSchedulePlanGuid(schedulePlanGuid);
+                    task.setTimeZone(context.getZone());
+                    task.setHealthCode(context.getHealthCode());
                     task.setActivity(activity);
-                    task.setScheduledOn(scheduledTime.getMillis());
+                    task.setLocalScheduledOn(scheduledTime.toLocalDateTime());
                     task.setGuid(BridgeUtils.generateGuid());
                     task.setPersistent(activity.isPersistentlyRescheduledBy(schedule));
                     if (expiresOn != null) {
-                        task.setExpiresOn(expiresOn.getMillis());
+                        task.setLocalExpiresOn(expiresOn.toLocalDateTime());
                         task.setHidesOn(expiresOn.getMillis());
                     }
-                    task.setRunKey(BridgeUtils.generateTaskRunKey(task));
+                    if (context.getSchedulePlanGuid() != null) {
+                        task.setRunKey(BridgeUtils.generateTaskRunKey(task, context));
+                    }
                     tasks.add(task);
                 }
             }
         }
     }
     
-    protected List<Task> trimTasks(List<Task> tasks) {
+    protected List<Task> trimTasks(List<Task> tasks, Schedule schedule) {
```

Comment:
Do we need to add schedule to the method signature? It's a member variable, and it doesn't look like anyone is overriding it.